File: content/snippets/cpp/interview_lcci_01_02_check_permutation.cpp
#include <array>
#include <iostream>
#include <string>

bool checkPermutation(const std::string& a, const std::string& b) {
    if (a.size() != b.size()) {
        return false;
    }

    std::array<int, 256> freq{};
    for (unsigned char ch : a) {
        ++freq[ch];
    }
    for (unsigned char ch : b) {
        --freq[ch];
        if (freq[ch] < 0) {
            return false;
        }
    }
    return true;
}

int main() {
    std::cout << "abcde vs edcba -> " << std::boolalpha << checkPermutation("abcde", "edcba") << '\n';
    std::cout << "abc vs abz -> " << std::boolalpha << checkPermutation("abc", "abz") << '\n';
    return 0;
}

File: content/snippets/typescript/interview_lcci_01_02_check_permutation.ts
function checkPermutation(a: string, b: string): boolean {
  if (a.length !== b.length) {
    return false;
  }

  const freq = new Map<string, number>();
  for (const ch of a) {
    freq.set(ch, (freq.get(ch) ?? 0) + 1);
  }

  for (const ch of b) {
    const left = (freq.get(ch) ?? 0) - 1;
    if (left < 0) {
      return false;
    }
    freq.set(ch, left);
  }

  return true;
}

function main(): void {
  console.log(`abcde vs edcba -> ${checkPermutation("abcde", "edcba")}`);
  console.log(`abc vs abz -> ${checkPermutation("abc", "abz")}`);
}

main();
